webhooks/services.py

```diff
@@ -1,14 +1,113 @@
+        return to_credentential_hook_model(item=item)
+
+    def _version_connections(self, item: dict) -> dict:
+        return to_connections_model(item=item)
+
+    def _deserialise(self, data: bytes) -> Union[str, dict]:
+        # Turn bytes from redis into str
+        try:
+            return (
+                data.decode("utf-8")
+                .replace("'", '"')
+                .replace("False", json.dumps(False))
+                .replace("True", json.dumps(True))
+            )
+        except JSONDecodeError:
+            # Don't raise an error here. It's probably more valuable to
+            # Return the data anyway than breaking the retrieval
+            return {"invalid JSON data": data}
+
+    def _to_json(self, data: List[bytes]) -> List[dict]:
+        data = [self._deserialise(val) for val in data]
+        try:
+            data = [json.loads(d) for d in data]
+        except JSONDecodeError:
+            # A multiline string has occured that has been split in
+            # an invalid way wrt JSON format
+
+            # Remove multiple spaces which occur when error msg is split into multiple lines
+            # turn off black formatter here because it replaces the regex
+            # fmt: off
+            data = [re.sub('\s+',' ', d) for d in data]
+            # Remove [..\" \"..] from the multiline error message because that is invalid json
+            data = [d.replace('\" \"', ' ') for d in data]
+            # fmt: on
+            data = [json.loads(d) for d in data]
+        return data
+
+    def _to_item(self, topic: str, data: List[dict]) -> List[dict]:
+        # Transform the data to the appropriate model
+        data = self._to_json(data=data)
+        if self._is_proof(topic):
+            data = [self._proof_hook_versioned(d) for d in data]
+        elif self._is_credential(topic):
+            data = [self._credential_hook_versioned(d) for d in data]
+        elif topic == "connections":
+            data = [self._version_connections(d) for d in data]
+        elif topic == "basicmessages":
+            data = [BasicMessagesHook(**d) for d in data]
+        return data
+
+    def _to_topic_item(self, topic: str, data: List[bytes]) -> List[TopicItem]:
+        data = self._to_item(topic=topic, data=data)
+        return (
+            [TopicItem(topic=topic, payload=payload) for payload in data]
+            if data
+            else []
+        )
+
+    async def add_topic_entry(self, topic: str, hook: bytes) -> List[TopicItem]:
         return await self._redis.sadd(topic, hook)
 
-    async def get_all_by_topic(self, topic: str) -> str:
-        return await self._redis.smembers(topic)
+    async def get_all_by_topic(self, topic: str) -> List[TopicItem]:
+        # Get the data from redis queue
+        data = await self._redis.smembers(topic)
+        # Transform it to TopicItems
+        items = self._to_topic_item(topic=topic, data=data)
+        return items
```

Comment:
Shouldn't we include paging here? What if we have a loooot of webhooks events?

Reply:
could add paging indeed. Need to have a look how to do that

Reply:
I've had a look and that is not a standard feature in fastapi so I have to use some plugin or roll my own

Reply:
since this will be accessed by code I think paging for human-readable friendliness is not required